include/beast/websocket/stream.hpp

```diff
@@ -627,6 +627,14 @@ class stream
         return wr_opcode_ == detail::opcode::text;
     }
 
+
+    /// Returns `true` if either the server or the client sent a close frame.
+    bool 
+    closed() const
+    {
+        return wr_close_ || rd_close_;
```

Comment:
You shouldn't be calling `async_close` twice.  If there is a scenario you can figure out that causes `async_close` to assert even when you only call it once, then that's a bug and needs to be fixed. The only case I can think of is if the implementation receives a close frame, sends one in response, and then before it delivers `websocket::error::close`, the application calls `async_close`. I can't prove that its possible though.

Now that I look over the async code it appears it is missing the assert present in the synchronous code, i will open an issue.

And I will refactor the close code to be more robust.